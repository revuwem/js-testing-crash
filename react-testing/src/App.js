import React from "react";
import { Routes, Route, Link } from "react-router-dom";
import Home from "./components/pages/Home";
import About from "./components/pages/About";
import Dashboard from "./components/pages/Dashboard";
import NotFound from "./components/pages/NotFound";

const App = () => {
  return (
    <div>
      <ul>
        <li>
          <Link to="/" data-testid="link-home">
            Home
          </Link>
        </li>
        <li>
          <Link to="/about" data-testid="link-about">
            About
          </Link>
        </li>
        <li>
          <Link to="/dashboard" data-testid="link-dashboard">
            Dashboard
          </Link>
        </li>
        <li>
          <Link to="/users" data-testid="link-users">
            Users
          </Link>
        </li>
      </ul>

      <hr />

      <Routes>
        <Route exact path="/" element={<Home />} />
        <Route path="/about" element={<About />} />
        <Route path="/dashboard" element={<Dashboard />} />
        <Route path="*" element={<NotFound />} />
      </Routes>
    </div>
  );
};

export default App;
