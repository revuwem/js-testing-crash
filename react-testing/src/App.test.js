import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";

import App from "./App";

describe("App", () => {
  test("should render all elements", () => {
    render(<App />);
    const helloWorldElement = screen.getByText(/hello world/i);
    expect(helloWorldElement).toBeInTheDocument();
    const button = screen.getByRole("button");
    expect(button).toBeInTheDocument();
    const input = screen.getByPlaceholderText(/type something/i);
    expect(input).toBeInTheDocument();
    // screen.debug();
    expect(input).toMatchSnapshot();
  });

  test("should not render 'hello2'", () => {
    render(<App />);
    const unwantedElement = screen.queryByText(/hello world2/i);
    expect(unwantedElement).toBeNull();
  });

  test("should render data container", async () => {
    render(<App />);
    const dataContainer = await screen.findByText(/data/i);
    expect(dataContainer).toBeInTheDocument();
    expect(dataContainer).toHaveStyle({ color: "red" });
  });

  test("toggle button should toggle", () => {
    render(<App />);
    const toggleBtn = screen.getByTestId("toggle-btn");
    expect(screen.queryByTestId("toggle-info")).toBeNull();
    fireEvent.click(toggleBtn);
    expect(screen.queryByTestId("toggle-info")).toBeInTheDocument();
    fireEvent.click(toggleBtn);
    expect(screen.queryByTestId("toggle-info")).toBeNull();
  });

  test("input value should change", () => {
    render(<App />);
    const inputValue = "something";
    const input = screen.getByPlaceholderText(/type something\.\.\./i);
    expect(screen.queryByTestId("input-value")).toContainHTML("");

    // fires explicit event
    // fireEvent.input(input, { target: { value: "something" } });

    // imitates user behaviour
    userEvent.type(input, inputValue);
    expect(screen.queryByTestId("input-value")).toContainHTML(inputValue);
  });
});
