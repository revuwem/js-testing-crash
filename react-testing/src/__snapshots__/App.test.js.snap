// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render all elements 1`] = `
<input
  id=""
  name=""
  placeholder="Type something..."
  type="text"
/>
`;
